### PR TITLE
Add OLE docs.

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

**Comment by nomis52** [comment]
@mrpackethead

**Comment by nomis52** [comment]
Merging now so we have some docs. I'll address comments in the next PR.

File: doxygen/main.md [hunk 1]
```diff
@@ -24,7 +24,7 @@ DMX receiver / RDM responder.
 
 ## DMX512 / RDM Receiver {#main-features-responder}
 
-- The ability to simulate many different types of @ref rdm-models.
+- The ability to simulate many different types of rdm-models.
```
Comment (peternewman) [inline doxygen/main.md]: RDM models (now we're not replacing it by the ref)

File: user_manual/docs/controller.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+[TOC]
+
+# Overview {#controller-overview}
+
+The device will automatically switch to controller mode when it's configured as
+an output port within OLA.
+
+When operating in controller mode, remember to set the biasing switches
+according to the manufacturer's instructions.
```
Comment (peternewman) [inline user_manual/docs/controller.md]: I know you've discussed this elsewhere, I think we should actually specify what biasing is required, and say see the manufacturer instructions for how to use it. As the termination should be the same regardless of who made it, it's just how to do it that changes.

Comment (peternewman) [inline user_manual/docs/controller.md]: Also we don't mention removing it when in responder mode.

File: user_manual/docs/firmware.md [hunk 3]
```diff
@@ -5,39 +5,47 @@ Updating Firmware
 
 # Overview {#update-overview}
 
-The **Firmware** is the code that runs on the board.
+The **Firmware** is the OLE code that runs on the device during normal operating
+conditions.
 
-The **Bootloader** is a mode that allows the firmware to be written to the board
+The **Bootloader** is a mode that allows the firmware to be written to the
```
Comment (peternewman) [inline user_manual/docs/firmware.md]: And the code that enables it. Perhaps try and be more consistent with the above.

File: user_manual/docs/firmware.md [hunk 4]
```diff
@@ -5,39 +5,47 @@ Updating Firmware
...
+device.
 
-# Application Firmware Update {#update-steps}
+The bootloader uses the 
+[Device Firmware Upgrade (DFU)]
```
Comment (peternewman) [inline user_manual/docs/firmware.md]: Wikipedia page, given it's the user manual?

Comment (peternewman) [inline user_manual/docs/firmware.md]: https://en.wikipedia.org/wiki/USB#DFU

File: user_manual/docs/firmware.md [hunk 5]
```diff
@@ -5,39 +5,47 @@ Updating Firmware
...
+protocol for upgrading firmware over USB. Any software that allows DFU uploads
+should work correctly with the device.
 
-## Enter Bootloader Mode {#update-bootloader}
-To enter the Bootloader mode press and hold the button on the board when powering up. If you are successful the top LED will be solid and the bottom three LEDS will flash together
+# Application Firmware Update {#update-steps}
 
 ## Flash Firmware {#update-flash}
-Download the latest firmware package from [github] and note its location
+Download the latest firmware package from [github] and note its location.
```
Comment (peternewman) [inline user_manual/docs/firmware.md]: SPaG GitHub
We also need to add an actual firmware release to there!

File: user_manual/docs/firmware.md [hunk 6]
```diff
@@ -5,39 +5,47 @@ Updating Firmware
...
-`sudo dfu-util -a 0 -R -D firmware.dfu`
+    $ sudo dfu-util -a 0 -R -D firmware.dfu
 
 ### On Mac {#update-flash-mac}
 
+The same instructions as Linux.
+
 ### On Windows {#update-flash-windows}
 
-# Bootloader Update {#bootloader-update}
```
Comment (peternewman) [inline user_manual/docs/firmware.md]: Should we add this to the developer manual?

File: user_manual/docs/host.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+[USB Console](@ref usb-console), the device must be connected to a host PC
+using a USB cable.
+
+# Drivers {#host-drivers}
+
+The device enumerates as a [CDC](https://en.wikipedia.org/wiki/USB_communications_device_class)
+and as a custom device class. The former is used for the USB Console while the
+latter is used for DMX / RDM controller messages.
+
+To use the device as a DMX/RDM controller, [libftdi](https://www.intra2net.com/en/developer/libftdi/)
```
Comment (peternewman) [inline user_manual/docs/host.md]: I'm sure you mean libusb

File: user_manual/docs/main.md [hunk 8]
```diff
@@ -3,10 +3,56 @@ Introduction     {#mainpage}
 
 [TOC]
 
-Explain what it is here.
+This guide describes how to get started with a device running the
+[Open Lighting Embedded](https://www.openlighting.org/ole/) (OLE) software. For
+developer information, including instructions on how to modify the OLE code,
+see the [OLE Developer Guide](http://docs.openlighting.org/ja-rule/doc/latest/).
```
Comment (peternewman) [inline user_manual/docs/main.md]: Change to /ole/

File: user_manual/docs/main.md [hunk 9]
```diff
@@ -3,10 +3,56 @@ Introduction     {#mainpage}
...
 
-Explain the relationship between the code and the device (maybe punt this to
-later since people will just want to get it working).
+Since the OLE software is customizable, the version running on a particular
+device may have different functionality from what is described here. Consult
+your manufacturer's documentation on the specific features enabled on a
+particular hardware product.
```
Comment (peternewman) [inline user_manual/docs/main.md]: and the code running on it...

File: user_manual/docs/main.md [hunk 10]
```diff
@@ -3,10 +3,56 @@ Introduction     {#mainpage}
...
+The examples in this guide are from a
+[Number1](https://www.openlighting.org/ole/number1/) since that was the first
+and most widely available OLE device. Throughout the rest of this document, the
+device running OLE is simply referred to as the 'device'.
 
+
+# Overview {#main-overview}
+
+The device can function as either an
+[RDM](https://en.wikipedia.org/wiki/RDM_(lighting%29) Responder or DMX / RDM
```
Comment (peternewman) [inline user_manual/docs/main.md]: Would it be worth being consistent and URL encoding both brackets?

File: user_manual/docs/main.md [hunk 11]
```diff
@@ -3,10 +3,56 @@ Introduction     {#mainpage}
...
+In controller mode, the device operates as a
+[USB to DMX / RDM Controller](@ref controller). This mode requires a host computer running the
+[Open Lighting Architecture](https://www.openlighting.org/ola/) (OLA). The OLA
+documentation outlines the requirements for the host computer.
+
+When the device is configured as an output port within OLA it will
+automatically switch into controller mode.
+
+When operating in controller mode, the OLE device reports detailed timing
+information about the RDM responses. The timing stats are reported when runinng
```
Comment (peternewman) [inline user_manual/docs/main.md]: SPaG running

File: user_manual/docs/rdm-responder.md [hunk 12]
```diff
@@ -1,6 +1,159 @@
-RDM Responder Mode
+RDM Responder  {#responder}
 ================
 
 [TOC]
 
 # Overview {#responder-overview}
+
+The OLE software can simulate different types of RDM Device Models. This
+allows RDM controllers to be tested without having to purchase a many
```
Comment (peternewman) [inline user_manual/docs/rdm-responder.md]: SPaG s/a //

File: user_manual/docs/rdm-responder.md [hunk 13]
```diff
@@ -1,6 +1,159 @@
...
+
+The RDM Device Model is controlled by Manufacturer Specific RDM PIDs. They can
+be changed with any RDM Controller software that supports setting Manufacturer
+Specific PIDs. These PIDs do not appear in SUPPORTED_PARAMETERS, so in a sense
+they are 'hidden'.
+
+### MODEL_ID_LIST
+
+[MODEL_ID_LIST](http://rdm.openlighting.org/pid/display?manufacturer=31344&pid=32771)
+will return a list of Device Model IDs. The request PDL is 0.
```
Comment (peternewman) [inline user_manual/docs/rdm-responder.md]: PDL is a bit advanced for users, I'd remove it, or word it that you can only get or something.

File: user_manual/docs/rdm-responder.md [hunk 14]
```diff
@@ -1,6 +1,159 @@
...
+[MODEL_ID](http://rdm.openlighting.org/pid/display?manufacturer=31344&pid=32770)
+is used to Get or Set the current Device Model ID.
+
+### OLA Commands
+
+The PIDs aren't included in the 0.10.0 release of OLA, so you will need to
+add [this file](https://raw.githubusercontent.com/OpenLightingProject/ja-rule/master/data/rdm/stellascapes_lightwidgets_number1.proto)
+to your PID data directory (normally /usr/local/share/ola/pids or /usr/share/ola/pids) and restart olad.
+
+Once you know the UID of your device. you can use  OLA CLI, to get a list of
```
Comment (peternewman) [inline user_manual/docs/rdm-responder.md]: I know their my typos, but there are two spaces before OLA, also I wonder if we should change CLI to command line tools.

File: user_manual/docs/rdm-responder.md [hunk 15]
```diff
@@ -1,6 +1,159 @@
...
+available models:
+
+    $ ola_rdm_get --universe 1 --uid 7a70:fffffe00 model_id_list
+
+To change the current model:
+
+    $ ola_rdm_set --universe 1 --uid 7a70:fffffe00 model_id 258
```
Comment (peternewman) [inline user_manual/docs/rdm-responder.md]: I take it setting a hex value works too, in which case that might be better so the docs tie up.

File: user_manual/docs/rdm-responder.md [hunk 16]
```diff
@@ -1,6 +1,159 @@
...
+
+# LED Model {#responder-led}
+
+The LED Model can be used to control SPI LED Pixels.
```
Comment (peternewman) [inline user_manual/docs/rdm-responder.md]: Connected how/where?

File: user_manual/docs/rdm-responder.md [hunk 17]
```diff
@@ -1,6 +1,159 @@
...
+--------------------------- | ----|-----|
+COMMS_STATUS                |  Y  |     |
+SUPPORTED_PARAMETERS        |  Y  |     |
+DEVICE_INFO                 |  Y  |     |
+PRODUCT_DETAIL_ID_LIST      |  Y  |     |
+DEVICE_MODEL_DESCRIPTION    |  Y  |     |
+BOOT_SOFTWARE_VERSION_LABEL |  Y  |     |
+DMX_PERSONALITY_DESCRIPTION |  Y  |     |
+
+and more..
```
Comment (peternewman) [inline user_manual/docs/rdm-responder.md]: Shouldn't we list all or nothing?

Should we link to the RDM website for each PID?

File: user_manual/docs/rdm-responder.md [hunk 18]
```diff
@@ -1,6 +1,159 @@
...
+
+# Sensor Model {#responder-sensor}
+
+This model has a DMX footprint of 0.
+
+Sensor Index | Name           | Type, Units                     | Units   | Notes
+-------------|----------------|---------------------------------|---------|---------
+0            | Temperature    | Temperature, Centigrade         | Deci    | Data is from the onboard temp. sensor if present.
+1            | Missing Sensor | Acceleration, meters / second^2 | Deci    | Always NACKs SENSOR_VALUE with NR_HARDWARE_FAULT
+2            | Voltage        | Voltage, Volts                  | Milli   | -
```
Comment (peternewman) [inline user_manual/docs/rdm-responder.md]: Is this a real value or a made up one?

File: user_manual/docs/rdm-responder.md [hunk 19]
```diff
@@ -1,6 +1,159 @@
...
+there are 4 sub-devices present.
+
+All PIDs from E1.37-1, Dimmer Message Sets are supported.
+
+## Sub-devices
+
+The dimmer has multiple sub-devices, that each take a single slot of DMX
+data. The sub-devices are not contiguous.
+
+DMX_BLOCK_ADDRESS can be used to set the DMX Start Address of all sub-devices in
```
Comment (peternewman) [inline user_manual/docs/rdm-responder.md]: Again, link to the RDM website?

**Comment by peternewman** [comment]
Some comments, mostly fairly minor.

You've somehow stopped it showing "Latest Git" as the version, unless I cheated before.

File: doxygen/main.md (hunk 1 above)
Comment (nomis52) [inline doxygen/main.md]: Done

File: user_manual/docs/controller.md (hunk 2 above)
Comment (nomis52) [inline user_manual/docs/controller.md]: Done. @mrpackethead  was going to provide a picture.

File: user_manual/docs/firmware.md (hunk 3 above)
Comment (nomis52) [inline user_manual/docs/firmware.md]: That's probably over the head of most users - they don't think of Bootloaders as 'code' .

I've changed it a bit to try to make it clear.

File: user_manual/docs/firmware.md (hunk 4 above)
Comment (nomis52) [inline user_manual/docs/firmware.md]: Done

File: user_manual/docs/firmware.md (hunk 5 above)
Comment (nomis52) [inline user_manual/docs/firmware.md]: Done. I'll upload some f/w

File: user_manual/docs/firmware.md (hunk 6 above)
Comment (nomis52) [inline user_manual/docs/firmware.md]: Done

File: user_manual/docs/host.md (hunk 7 above)
Comment (nomis52) [inline user_manual/docs/host.md]: Yup, was not thinking here.

File: user_manual/docs/main.md (hunk 8 above)
Comment (nomis52) [inline user_manual/docs/main.md]: Done

File: user_manual/docs/main.md (hunk 9 above)
Comment (nomis52) [inline user_manual/docs/main.md]: I think 'code' is too technical here. What are we trying to say?

File: user_manual/docs/main.md (hunk 10 above)
Comment (nomis52) [inline user_manual/docs/main.md]: Done

File: user_manual/docs/main.md (hunk 11 above)
Comment (nomis52) [inline user_manual/docs/main.md]: Done

File: user_manual/docs/rdm-responder.md (hunk 12 above)
Comment (nomis52) [inline user_manual/docs/rdm-responder.md]: Done

File: user_manual/docs/rdm-responder.md (hunk 13 above)
Comment (nomis52) [inline user_manual/docs/rdm-responder.md]: Done

File: user_manual/docs/rdm-responder.md (hunk 14 above)
Comment (nomis52) [inline user_manual/docs/rdm-responder.md]: Done

File: user_manual/docs/rdm-responder.md (hunk 16 above)
Comment (nomis52) [inline user_manual/docs/rdm-responder.md]: I'll be filling that in later.

File: user_manual/docs/rdm-responder.md (hunk 15 above)
Comment (nomis52) [inline user_manual/docs/rdm-responder.md]: Well the numbers are decimal when the get command is displayed. I've added the decimal values to the table above.

File: user_manual/docs/rdm-responder.md (hunk 17 above)
Comment (nomis52) [inline user_manual/docs/rdm-responder.md]: I'm writing a script to generate these. Yes, there will be links.

File: user_manual/docs/rdm-responder.md (hunk 18 above)
Comment (nomis52) [inline user_manual/docs/rdm-responder.md]: Added a note.

File: user_manual/docs/rdm-responder.md (hunk 19 above)
Comment (nomis52) [inline user_manual/docs/rdm-responder.md]: Done

File: user_manual/docs/main.md (hunk 9 above)
Comment (peternewman) [inline user_manual/docs/main.md]: Software perhaps then. You first say the software is customisable, then you say ask your manufacturer about what your particular hardware enables. They may have a Number 1 with custom firmware on it, loaded by someone else.

File: user_manual/docs/rdm-responder.md (hunk 15 above)
Comment (peternewman) [inline user_manual/docs/rdm-responder.md]: That'll do.

File: user_manual/docs/rdm-responder.md (hunk 17 above)
Comment (peternewman) [inline user_manual/docs/rdm-responder.md]: Heh, funnily enough I was thinking the same thing.

File: user_manual/docs/rdm-responder.md (hunk 19 above)
Comment (peternewman) [inline user_manual/docs/rdm-responder.md]: I'd be inclined to do this for all the other times it occurs too. Thoughts?

File: user_manual/docs/main.md (hunk 9 above)
Comment (nomis52) [inline user_manual/docs/main.md]: I've removed hardware.

File: user_manual/docs/rdm-responder.md (hunk 19 above)
Comment (nomis52) [inline user_manual/docs/rdm-responder.md]: Too much work.

File: user_manual/docs/main.md (hunk 9 above)
Comment (peternewman) [inline user_manual/docs/main.md]: Cool.

File: user_manual/docs/rdm-responder.md (hunk 19 above)
Comment (peternewman) [inline user_manual/docs/rdm-responder.md]: Okay, I might look at adapting your code to script it.

**Pull request merged**